contracts/btc-finality/src/finality.rs

```diff
@@ -622,3 +627,35 @@ pub fn list_fps_by_power(
     let res: FinalityProvidersByPowerResponse = querier.query(&query)?;
     Ok(res.fps)
 }
+
+/// `distribute_rewards` distributes rewards to finality providers who have voted for a block at `height`
+pub fn distribute_rewards(deps: &mut DepsMut, height: u64) -> Result<(), ContractError> {
+    let voting_fps = VOTES
+        .prefix(height)
+        .range(deps.storage, None, None, Ascending)
+        .collect::<StdResult<Vec<_>>>()?;
+    // Get the total voting power of the voting FPS
+    let total_voting_power = voting_fps
+        .iter()
+        .map(|(_, vote)| vote.voting_power as u128)
+        .sum::<u128>();
+    // Get the rewards to distribute (bank balance of the staking contract)
+    let cfg = CONFIG.load(deps.storage)?;
+    let rewards_amount = deps.querier.query_balance(cfg.staking, cfg.denom)?.amount;
```

Comment:
Since the protocol generates all rewards upon finalising blocks, would the following situation happen:

- Suddenly the protocol finalises 10 blocks and generates rewards for these blocks
- The protocol distributes the rewards of these 10 blocks to the FPs *in the last block*
- The FPs for the rest 9 blocks do not get reward

Babylon resolves this problem by having a [KV store](https://github.com/babylonlabs-io/babylon/blob/6f77345f0242765e5de3c5df261a431ebafe56ae/x/incentive/keeper/btc_staking_gauge.go#L79-L88) to track the reward for every height, and the protocol distributes rewards in each height's gauge to the FPs at that height. Maybe we need something similar?

Reply:
I tried to avoid doing that. See #97 description for rationale.